docs/RFCs/19042021_self_signed_certificates.md

```diff
@@ -0,0 +1,240 @@
+ 
+#### Check-cert-for-regeneration
+* Check if the certificate secret has annotation `managed-by: cockroachdb`, if not return False
+* Check if the certificate is expiring in the next 2 months using the values from annotations `creationTime` and `duration` on secret, if expiring return True, else False
+ 
+#### Validate-Secret-Annotations
+* Check if annotation for `resourceVersion` and `creationTime` exists, if not return False
+* Check if resourceVersion of the secret matches with the value of `resourceVersion` annotation, if not means secret is changed, return False
+* Else, return True
+ 
+### Certificate Generation cases during Helm upgrade:
+ 
+In case of Helm upgrade:
+ 
+* User has given CA and changes contents of the CA secret:
+  * Check if the current value `resourceVersion` or hash matches with the annotation value. if annotation does not match, so this is a new CA scenario.
+* User has given CA and changes the secret name:
+  * Annotation will not be found, so this is a new CA scenario.
+* User had not given CA previously, but now has given the CA:
+  * Annotation will not be found, so this is a new CA scenario
+* If the user changes the duration of CA:
+  * Identify and compare using existing annotation on CA secret and current value, this will be a case for certificate rotation. This will only be the case when the CA is managed by CockroachDB.
+* User changes the duration of all certificates:
+  * Compare old and new CA duration from CA secret annotation values and current value. This will be a case for certificate rotation.
+  * Rotate CA certificate and add an annotation on CA secret with the date of rotation.
+  * Add annotations on Node and Client certs specifying the new expected duration and `to-be-rotated: true`.
+  * These secret certificates will be renewed in the next cron cycle and `to-be-rotated:true` and expected duration annotations will be removed.
+* User only changes the duration of either Node or Client certificate:
+  * Identify and compare duration with existing annotation value and current value and renew Node or Client certificate.
+* User certificate management method is changed from certificate generation to `cert-manager` or `default manual k8s CSR approval`:
+  * Do nothing as this `pre-install` job  won't be triggered.
+ 
+### Periodic Rotation scenarios:
+* CA certificate is near expiry: This will be identified using the generation time put on the CA secret. This will lead to a certificate rotation scenario.
+* Node or Client certificate near expiry: This will be identified using the generation time put on the respective secret. This will lead to regeneration of Node or Client certificate
+ 
+### Certificate Rotation scenario:
+ 
+* Only renew CA certificate by generating new CA key and new combined CA along with the old CA.
+* Update CA secret and add CA certificate info in Node and Client secret.
+`TODO: Need to identify how to generate the SIGHUP signal in all the Nodes for certificate renewal`
```

Comment:
This todo is at 3 places in this doc

Reply:
This is added wherever update is required.